File: Editorial/1836_Remove_Duplicates_From_an_Unsorted_Linked_List/Python3/2_Recursive_Hash_Map.py
class Solution:
    def deleteDuplicatesUnsorted(self, head: ListNode) -> ListNode:
        frequency = {}
        self.count_frequencies(head, frequency)
        return self.delete_duplicates_unsorted_helper(head, frequency)

    # Count the frequency of each value in the list
    def count_frequencies(self, head: ListNode, frequency: dict):
        current = head
        while current is not None:
            frequency[current.val] = frequency.get(current.val, 0) + 1
            current = current.next

    # Recursively delete duplicates based on the frequency map
    def delete_duplicates_unsorted_helper(
        self, head: ListNode, frequency: dict
    ) -> ListNode:
        if head is None:
            return None

        # Recursive call for the next node
        updated_next_node = self.delete_duplicates_unsorted_helper(
            head.next, frequency
        )
        head.next = updated_next_node

        # If the current node is a duplicate, return the updated next node
        if frequency[head.val] > 1:
            return updated_next_node

        # Otherwise, return the current node
        return head

File: Editorial/1836_Remove_Duplicates_From_an_Unsorted_Linked_List/CPP/2_Recursive_Hash_Map.cpp
class Solution {
public:
    ListNode* deleteDuplicatesUnsorted(ListNode* head) {
        unordered_map<int, int> frequency;
        countFrequencies(head, frequency);
        return deleteDuplicatesUnsortedHelper(head, frequency);
    }

private:
    // Count the frequency of each value in the list
    void countFrequencies(ListNode* head, unordered_map<int, int>& frequency) {
        ListNode* current = head;
        while (current != nullptr) {
            ++frequency[current->val];
            current = current->next;
        }
    }

    // Recursively delete duplicates based on the frequency map
    ListNode* deleteDuplicatesUnsortedHelper(
        ListNode* head, unordered_map<int, int>& frequency) {
        if (!head) {
            return nullptr;
        }

        // Recursive call for the next node
        ListNode* updatedNextNode =
            deleteDuplicatesUnsortedHelper(head->next, frequency);
        head->next = updatedNextNode;

        // If the current node is a duplicate, return the updated next node
        if (frequency[head->val] > 1) {
            return updatedNextNode;
        }

        // Otherwise, return the current node
        return head;
    }
};
